svsim/src/test/scala/BackendSpec.scala

```diff
@@ -298,6 +298,30 @@ trait BackendSpec extends AnyFunSpec with Matchers {
           }
         }
       }
+
+      it("ends the simulation on '$finish' (#4700)") {
+        workspace.reset()
+        workspace.elaborateFinishTest()
+        workspace.generateAdditionalSources()
+        simulation = workspace.compile(
+          backend
+        )(
+          workingDirectoryTag = name,
+          commonSettings = CommonCompilationSettings(),
+          backendSpecificSettings = compilationSettings,
+          customSimulationWorkingDirectory = None,
+          verbose = false
+        )
+        intercept[svsim.Simulation.UnexpectedEndOfMessages.type] {
+          simulation.run(
+            verbose = false,
+            executionScriptLimit = None
+          ) { controller =>
+            controller.run(1)
+            println(controller.readLog())
```

Comment:
Should this be checking something instead of just printing?

Reply:
I thought that this was testing it properly, but it was not. The exception was coming from the simulation shutting down and then the read would fail. This isn't the best test.

I changed that in the force push to count the number of finishes in the simulation log.